base_exception/models/base_exception.py

```diff
@@ -91,6 +92,22 @@ def _popup_exceptions(self):
     def _get_popup_action(self):
         return self.env.ref("base_exception.action_exception_rule_confirm")
 
+    def _add_detected_exceptions_to_self(self):
+        return self != self._get_main_records()
+
+    def _detect_exceptions(self, rule):
+        records = super()._detect_exceptions(rule)
+        # If _get_main_records returns self, it adds the exceptions to itself in
+        # detect_exceptions. If _get_main_records does not return self, it adds the
+        # exceptions here
+        if not self._add_detected_exceptions_to_self():
+            return records
+
+        (self - records).exception_ids = [(3, rule.id)]
+        records.exception_ids = [(4, rule.id)]
```

Comment:
Here a small minor request. 
```suggestion
        if self._add_detected_exceptions_to_self():
            (self - records).exception_ids = [(3, rule.id)]
            records.exception_ids = [(4, rule.id)]
```

Reply:
Is it correct that since you are writing the rules now on the records, the logic in `detect_exceptions` in `base.exception.method` is not used anymore. Or only if the returned main record isn't the same as where it was executed from?

Isn't this quite bad because the checks and writes in `detect_exceptions` are done twice?

Isn't it possible to have the writing in just one place?

Reply:
`detect_exceptions` in `base.exception.method` is still needed.
If a model inherits from `base.exception.method`, it has no `exception_ids`, that's why it must return another record in `_get_main_records`. The detected rules in `base.exception.method` are written to that other model. Like it is done right now in `account_move_exception`.

If a model inherits from `base.exception`, there are 2 ways:
1. It overrides `_get_main_records`, like it is done in my PR for `sale_exception`. But as it returns another record, it will not write the detected rules on the sale order lines, only on the sales orders returned by `_get_main_records`. That's why this override here is needed
2. It does not override `_get_main_records`: Then the detected rules are added to self in `base.exception.method` and this here has no effect as it is always False

Reply:
Ok. Since you are only adding it if the main record isn't the same. So if the exception is on sale.order and it is detected on sale.order it will add it in `detect_exceptions`.